File: sdks/dotnet/SutraError.cs
namespace SutraDB.Client;

/// <summary>
/// Exception thrown when a SutraDB operation fails.
/// </summary>
public class SutraException : Exception
{
    /// <summary>
    /// The HTTP status code returned by the server, or null if the error is not HTTP-related.
    /// </summary>
    public int? StatusCode { get; }

    /// <summary>
    /// Create a new SutraException with a message and optional status code.
    /// </summary>
    public SutraException(string message, int? statusCode = null)
        : base(message)
    {
        StatusCode = statusCode;
    }

    /// <summary>
    /// Create a new SutraException wrapping an inner exception.
    /// </summary>
    public SutraException(string message, Exception innerException)
        : base(message, innerException)
    {
        StatusCode = null;
    }
}

File: sdks/java/src/main/java/io/github/emmaleonhart/sutradb/SutraError.java
package io.github.emmaleonhart.sutradb;

/**
 * Exception thrown when a SutraDB operation fails.
 *
 * <p>Wraps both HTTP-level errors (connection failures, timeouts) and
 * application-level errors returned by the SutraDB server.</p>
 */
public class SutraError extends RuntimeException {

    private final int statusCode;

    /**
     * Create a new SutraError with a message and HTTP status code.
     *
     * @param message    human-readable error description
     * @param statusCode the HTTP status code, or -1 if the error is not HTTP-related
     */
    public SutraError(String message, int statusCode) {
        super(message);
        this.statusCode = statusCode;
    }

    /**
     * Create a new SutraError wrapping another throwable.
     *
     * @param message human-readable error description
     * @param cause   the underlying cause
     */
    public SutraError(String message, Throwable cause) {
        super(message, cause);
        this.statusCode = -1;
    }

    /**
     * Return the HTTP status code associated with this error,
     * or -1 if the error is not HTTP-related.
     *
     * @return HTTP status code or -1
     */
    public int getStatusCode() {
        return statusCode;
    }
}
